File: app/java/src/main/java/com/example/production/infrastructure/web/controller/PurchaseOrderController.java
package com.example.production.infrastructure.web.controller;

import com.example.production.application.port.in.PurchaseOrderUseCase;
import com.example.production.application.port.in.command.CreatePurchaseOrderCommand;
import com.example.production.domain.model.purchase.PurchaseOrder;
import com.example.production.infrastructure.web.dto.CreatePurchaseOrderRequest;
import com.example.production.infrastructure.web.dto.PurchaseOrderResponse;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * 発注 Controller（Input Adapter）
 */
@RestController
@RequestMapping("/api/purchase-orders")
@Tag(name = "purchase-orders", description = "発注 API")
@RequiredArgsConstructor
public class PurchaseOrderController {

    private final PurchaseOrderUseCase useCase;

    @GetMapping
    @Operation(summary = "発注一覧の取得")
    public ResponseEntity<List<PurchaseOrderResponse>> getAllOrders() {
        List<PurchaseOrder> orders = useCase.getAllOrders();
        return ResponseEntity.ok(orders.stream()
                .map(PurchaseOrderResponse::from)
                .toList());
    }

    @GetMapping("/{orderNumber}")
    @Operation(summary = "発注詳細の取得")
    public ResponseEntity<PurchaseOrderResponse> getOrder(
            @PathVariable String orderNumber) {
        PurchaseOrder order = useCase.getOrder(orderNumber);
        return ResponseEntity.ok(PurchaseOrderResponse.from(order));
    }

    @PostMapping
    @Operation(summary = "発注の登録")
    public ResponseEntity<PurchaseOrderResponse> createOrder(
            @Valid @RequestBody CreatePurchaseOrderRequest request) {

        CreatePurchaseOrderCommand command = CreatePurchaseOrderCommand.builder()
                .supplierCode(request.getSupplierCode())
                .ordererCode(request.getOrdererCode())
                .departmentCode(request.getDepartmentCode())
                .remarks(request.getRemarks())
                .details(request.getDetails().stream()
                        .map(d -> CreatePurchaseOrderCommand.PurchaseOrderDetailCommand.builder()
                                .itemCode(d.getItemCode())
                                .orderQuantity(d.getOrderQuantity())
                                .unitPrice(d.getUnitPrice())
                                .deliveryDate(d.getDeliveryDate())
                                .build())
                        .toList())
                .build();

        PurchaseOrder order = useCase.createOrder(command);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(PurchaseOrderResponse.from(order));
    }

    @PostMapping("/{orderNumber}/confirm")
    @Operation(summary = "発注の確定")
    public ResponseEntity<PurchaseOrderResponse> confirmOrder(
            @PathVariable String orderNumber) {
        PurchaseOrder order = useCase.confirmOrder(orderNumber);
        return ResponseEntity.ok(PurchaseOrderResponse.from(order));
    }

    @DeleteMapping("/{orderNumber}")
    @Operation(summary = "発注の取消")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void cancelOrder(@PathVariable String orderNumber) {
        useCase.cancelOrder(orderNumber);
    }
}

File: app/csharp/src/ProductionManagement.Infrastructure/Rest/Controllers/PurchaseOrderController.cs
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using ProductionManagement.Application.Port.In;
using ProductionManagement.Application.Port.In.Command;
using ProductionManagement.Infrastructure.Rest.Dto;

namespace ProductionManagement.Infrastructure.Rest.Controllers;

/// <summary>
/// 発注 Controller（Input Adapter）
/// </summary>
[ApiController]
[Route("api/purchase-orders")]
[Tags("Purchase Orders")]
public class PurchaseOrderController : ControllerBase
{
    private readonly IPurchaseOrderUseCase _purchaseOrderUseCase;

    public PurchaseOrderController(IPurchaseOrderUseCase purchaseOrderUseCase)
    {
        _purchaseOrderUseCase = purchaseOrderUseCase;
    }

    /// <summary>
    /// 発注一覧の取得
    /// </summary>
    [HttpGet]
    [ProducesResponseType(typeof(IReadOnlyList<PurchaseOrderResponse>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetAllOrders()
    {
        var orders = await _purchaseOrderUseCase.GetAllOrdersAsync();
        return Ok(orders.Select(PurchaseOrderResponse.FromWithoutDetails).ToList());
    }

    /// <summary>
    /// 発注の取得
    /// </summary>
    /// <param name="purchaseOrderNumber">発注番号</param>
    [HttpGet("{purchaseOrderNumber}")]
    [ProducesResponseType(typeof(PurchaseOrderResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> GetOrder(string purchaseOrderNumber)
    {
        var order = await _purchaseOrderUseCase.GetOrderAsync(purchaseOrderNumber);
        return Ok(PurchaseOrderResponse.From(order));
    }

    /// <summary>
    /// 発注の作成
    /// </summary>
    [HttpPost]
    [ProducesResponseType(typeof(PurchaseOrderResponse), StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<IActionResult> CreateOrder([FromBody] CreatePurchaseOrderRequest request)
    {
        var command = new PurchaseOrderCreateCommand
        {
            SupplierCode = request.SupplierCode,
            OrderDate = request.OrderDate,
            OrdererCode = request.OrdererCode,
            DepartmentCode = request.DepartmentCode,
            TaxRate = request.TaxRate,
            Remarks = request.Remarks,
            Details = request.Details.Select(d => new PurchaseOrderDetailCommand
            {
                ItemCode = d.ItemCode,
                OrderQuantity = d.OrderQuantity,
                ExpectedReceivingDate = d.ExpectedReceivingDate,
                OrderNumber = d.OrderNumber,
                DeliveryLocationCode = d.DeliveryLocationCode
            }).ToList()
        };

        var order = await _purchaseOrderUseCase.CreateOrderAsync(command);
        return CreatedAtAction(
            nameof(GetOrder),
            new { purchaseOrderNumber = order.PurchaseOrderNumber },
            PurchaseOrderResponse.From(order)
        );
    }

    /// <summary>
    /// 発注の確定
    /// </summary>
    /// <param name="purchaseOrderNumber">発注番号</param>
    [HttpPost("{purchaseOrderNumber}/confirm")]
    [ProducesResponseType(typeof(PurchaseOrderResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> ConfirmOrder(string purchaseOrderNumber)
    {
        var order = await _purchaseOrderUseCase.ConfirmOrderAsync(purchaseOrderNumber);
        return Ok(PurchaseOrderResponse.FromWithoutDetails(order));
    }

    /// <summary>
    /// 発注の取消
    /// </summary>
    /// <param name="purchaseOrderNumber">発注番号</param>
    [HttpPost("{purchaseOrderNumber}/cancel")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> CancelOrder(string purchaseOrderNumber)
    {
        await _purchaseOrderUseCase.CancelOrderAsync(purchaseOrderNumber);
        return NoContent();
    }
}
